In [1]:
import pandas as pd
import numpy as np

### Затягиваем датасет

In [182]:
df_penguins = pd.read_csv(r'https://raw.githubusercontent.com/YBI-Foundation/Dataset/main/Penguins.csv')

In [44]:
df_penguins.head(15)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0   Adelie  Torgersen            39.1           18.7              181.0   
1   Adelie  Torgersen            39.5           17.4              186.0   
2   Adelie  Torgersen            40.3           18.0              195.0   
3   Adelie  Torgersen             NaN            NaN                NaN   
4   Adelie  Torgersen            36.7           19.3              193.0   
5   Adelie  Torgersen            39.3           20.6              190.0   
6   Adelie  Torgersen            38.9           17.8              181.0   
7   Adelie  Torgersen            39.2           19.6              195.0   
8   Adelie  Torgersen            34.1           18.1              193.0   
9   Adelie  Torgersen            42.0           20.2              190.0   
10  Adelie  Torgersen            37.8           17.1              186.0   
11  Adelie  Torgersen            37.8           17.3              180.0   
12  Adelie  Torgersen            41.1           17.6              182.0   
13  Adelie  Torgersen            38.6           21.2              191.0   
14  Adelie  Torgersen            34.6           21.1              198.0   

    body_mass_g     sex  
0        3750.0    MALE  
1        3800.0  FEMALE  
2        3250.0  FEMALE  
3           NaN     NaN  
4        3450.0  FEMALE  
5        3650.0    MALE  
6        3625.0  FEMALE  
7        4675.0    MALE  
8        3475.0     NaN  
9        4250.0     NaN  
10       3300.0     NaN  
11       3700.0     NaN  
12       3200.0  FEMALE  
13       3800.0    MALE  
14       4400.0    MALE

In [180]:
df_penguins.tail(15)

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
329  Gentoo  Biscoe            48.1           15.1              209.0   
330  Gentoo  Biscoe            50.5           15.2              216.0   
331  Gentoo  Biscoe            49.8           15.9              229.0   
332  Gentoo  Biscoe            43.5           15.2              213.0   
333  Gentoo  Biscoe            51.5           16.3              230.0   
334  Gentoo  Biscoe            46.2           14.1              217.0   
335  Gentoo  Biscoe            55.1           16.0              230.0   
336  Gentoo  Biscoe            44.5           15.7              217.0   
337  Gentoo  Biscoe            48.8           16.2              222.0   
338  Gentoo  Biscoe            47.2           13.7              214.0   
339  Gentoo  Biscoe             NaN            NaN                NaN   
340  Gentoo  Biscoe            46.8           14.3              215.0   
341  Gentoo  Biscoe            50.4           15.7              222.0   
342  Gentoo  Biscoe            45.2           14.8              212.0   
343  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
329       4050.0    MALE  
330       4050.0  FEMALE  
331       4050.0    MALE  
332       4050.0  FEMALE  
333       4050.0    MALE  
334       4050.0  FEMALE  
335       4050.0    MALE  
336       4050.0     NaN  
337       4050.0    MALE  
338       4050.0  FEMALE  
339       4050.0     NaN  
340       4050.0  FEMALE  
341       4050.0    MALE  
342       4050.0  FEMALE  
343       4050.0    MALE

### Делим и собираем обратно конкатом

In [178]:
df_penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      342.000000     342.000000         342.000000        344.0
mean        43.921930      17.151170         200.915205       4050.0
std          5.459584       1.974793          14.061714          0.0
min         32.100000      13.100000         172.000000       4050.0
25%         39.225000      15.600000         190.000000       4050.0
50%         44.450000      17.300000         197.000000       4050.0
75%         48.500000      18.700000         213.000000       4050.0
max         59.600000      21.500000         231.000000       4050.0

### почему-то сумма двух частей (172 и 170) не дает 344 строки... не могу понять почему. Помогите пожалуйста

In [86]:
first_df = df_penguins[df_penguins['body_mass_g'] >= 4050.000000]
second_df = df_penguins[df_penguins['body_mass_g'] < 4050.000000]

In [106]:
a = df_penguins.shape
b = first_df.shape
c = second_df.shape
print (a, b, c)

(344, 7) (172, 7) (170, 7)


In [115]:
first_df.columns == second_df.columns

array([ True,  True,  True,  True,  True,  True,  True])

In [119]:
pd.concat([first_df, second_df]).sort_index() #получается собрал на 2 строки меньше... не могу разобраться что не так

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
4    Adelie  Torgersen            36.7           19.3              193.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
..      ...        ...             ...            ...                ...   
338  Gentoo     Biscoe            47.2           13.7              214.0   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
4         3450.0  FEMALE  
5         3650.0    MALE  
..           ...     ...  
338       4925.0  FEMALE  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[342 rows x 7 columns]

In [135]:
df_union = pd.concat([first_df, second_df]).sort_index()

In [176]:
df_union

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
4    Adelie  Torgersen            36.7           19.3              193.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
..      ...        ...             ...            ...                ...   
338  Gentoo     Biscoe            47.2           13.7              214.0   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
4         3450.0  FEMALE  
5         3650.0    MALE  
..           ...     ...  
338       4925.0  FEMALE  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[342 rows x 7 columns]

In [143]:
df_penguins.columns == df_union.columns

array([ True,  True,  True,  True,  True,  True,  True])

### пропуски

In [150]:
df_penguins.isna().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           0
sex                  11
dtype: int64

In [152]:
df_penguins.dropna()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
4    Adelie  Torgersen            36.7           19.3              193.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
..      ...        ...             ...            ...                ...   
338  Gentoo     Biscoe            47.2           13.7              214.0   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         4050.0    MALE  
1         4050.0  FEMALE  
2         4050.0  FEMALE  
4         4050.0  FEMALE  
5         4050.0    MALE  
..           ...     ...  
338       4050.0  FEMALE  
340       4050.0  FEMALE  
341       4050.0    MALE  
342       4050.0  FEMALE  
343       4050.0    MALE  

[333 rows x 7 columns]

In [154]:
df_penguins[df_penguins['sex'].isna()]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen             NaN            NaN                NaN   
8    Adelie  Torgersen            34.1           18.1              193.0   
9    Adelie  Torgersen            42.0           20.2              190.0   
10   Adelie  Torgersen            37.8           17.1              186.0   
11   Adelie  Torgersen            37.8           17.3              180.0   
47   Adelie      Dream            37.5           18.9              179.0   
246  Gentoo     Biscoe            44.5           14.3              216.0   
286  Gentoo     Biscoe            46.2           14.4              214.0   
324  Gentoo     Biscoe            47.3           13.8              216.0   
336  Gentoo     Biscoe            44.5           15.7              217.0   
339  Gentoo     Biscoe             NaN            NaN                NaN   

     body_mass_g  sex  
3         4050.0  NaN  
8         4050.0  NaN  
9         4050.0  NaN  
10        4050.0  NaN  
11        4050.0  NaN  
47        4050.0  NaN  
246       4050.0  NaN  
286       4050.0  NaN  
324       4050.0  NaN  
336       4050.0  NaN  
339       4050.0  NaN

In [166]:
df_penguins['sex'].fillna('NOT SPECIFIED')

0               MALE
1             FEMALE
2             FEMALE
3      NOT SPECIFIED
4             FEMALE
           ...      
339    NOT SPECIFIED
340           FEMALE
341             MALE
342           FEMALE
343             MALE
Name: sex, Length: 344, dtype: object

### Дубликаты

In [171]:
df_penguins[df_penguins.duplicated()] # я так понял дубликатов нет
#если бы были удалил бы так: df_penguins.drop_duplicates()

Empty DataFrame
Columns: [species, island, bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g, sex]
Index: []

### выбросы

In [184]:
mu = df_union['body_mass_g'].mean()
sigma = df_union['body_mass_g'].std()
lower_bound = mu - 3*sigma
upper_bound = mu + 3*sigma
print(f"Выбросами будут считаться все массы пингвинов, которые находится за пределами интервала [{lower_bound}, {upper_bound}]")

Выбросами будут считаться все массы пингвинов, которые находится за пределами интервала [1795.890778870626, 6607.617993059199]


In [186]:
df_no_outliers = df_union[(df_union['body_mass_g']>=lower_bound)&(df_union['body_mass_g']<=upper_bound)]

In [188]:
df_no_outliers

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
4    Adelie  Torgersen            36.7           19.3              193.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
..      ...        ...             ...            ...                ...   
338  Gentoo     Biscoe            47.2           13.7              214.0   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
4         3450.0  FEMALE  
5         3650.0    MALE  
..           ...     ...  
338       4925.0  FEMALE  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[342 rows x 7 columns]

### нормализация

In [191]:
df_no_outliers['norm_body_mass_g'] = (df_no_outliers['body_mass_g'] - df_no_outliers['body_mass_g'].min()) / (df_no_outliers['body_mass_g'].max() - df_no_outliers['body_mass_g'].min()) 

In [193]:
df_no_outliers['norm_body_mass_g'].describe()

count    342.000000
mean       0.417154
std        0.222765
min        0.000000
25%        0.236111
50%        0.375000
75%        0.569444
max        1.000000
Name: norm_body_mass_g, dtype: float64